### ISSUE TITLE
[LGSM] Steam Guard /2 Factor Auth still not working

### ISSUE THREAD

**Issue by DSDV** [dgibbs64/linuxgsm#579]
As reported here: https://github.com/dgibbs64/linuxgsm/issues/298

SteamGuard Two Factor auth seems to make some trouble.
As of now its implemented but not relay working as you can see on the folowing screenshot wich looked like these aover 5 hours... also tried updateing and so on.

http://imgur.com/R2pD4xJ

**Comment by DSDV** [comment]
PS seems to be an problem with the mobile APP while with another account wich sadly still uses eMail it seems to work for another game.

I think its because eMail verification is somehow stored and the APP must be entered every time you log in

**Comment by UltimateByte** [comment]
We should try this in standalone (without lgsm script) mode to see if srcds enven manages to get it working.

**Comment by DSDV** [comment]
since lsgm sadly dont work with games where i need to use steamguard protected accounts the way u discriebed is the only one left to use for me - and it works well except its a 'lot' more work.

**Comment by dgibbs64** [comment]
This issue is definitely something that needs to be sorted out. It is on my radar for working on.

**Comment by UltimateByte** [comment]
So it's fixable, that's a positive. But did you try without setting the steampass variable in the end ?

**Comment by DSDV** [comment]
Nope i only tried the script "as is" and dont changed something

**Comment by UltimateByte** [comment]
Try with the script, just don't set a password in the server config, and see what happens at install.

**Comment by dgibbs64** [comment]
Just tested with Steam Guard Authenticator. The issue is that it asks for a code every time you login to SteamCMD. 

If you want to do any automation of server updates this is an issue straight away as user intervention is requires each time LGSM uses SteamCMD.

LGSM uses the tee command to send the output of SteamCMD to a log file and this is why it sometimes looks like its doing nothing on update (to do with a buffer or somthing). 

My original issue was that Steam Guard Email code entry was not being seen by the user because of tee. 

To get around this issue LGSM runs ./steamcmd.sh +login +password +quir without tee before the update command. This will ensure that if a code is required is is seen by the user, preventing SteamCMD from looking stuck when its waiting for user input.


I had setup SteamCMD to just login and quit without tee to allow Steam Guard Email codes to be input. These only need to be input once on each server.

So I recommend either using Steam Guard Email over Steam Guard Authenticator. Or purchasing the game on another steam account just for your servers.

Its a pain but I dont think Steam Guard Authenticator will work with LGSM and will break updating servers automatically.

I will figure out the best way to notify the user of this to ensure they are not using Steam Guard Authenticator.

**Comment by DSDV** [comment]
Hm.. ok its what it what i was afraid of.

But at least we have a workaround.


2 Thoughts on this: 
I think they will push the Mobile version harder in future and will shut down eMail eventualy.
So can we implement an authenticator in LSGM ? (i now securitywise its rubbish but still)
I know ther is a solution for windows (WinAuth) to get codes on WIndows based platforms (wich are not supported)


with best (and sorry for my bad english) +DS_DV+

**Comment by dgibbs64** [comment]
@DSDV  I think you miss the point that the Authenticator is not suitable for this kind of setup. I don't think that Valve will phase out email authentication at all. Not all users will want to/cannot use a smartphone app for authenticating there steam account. Steam Guard via email is better than no Steam Guard but not quite as good as via smartphone. 

Automation is simply not possible with the Authenticator as it requires user interaction for every login. It is not feasible for servers as they can receive updates at any point, and login to steam several times a day.

Steam Guard via email works by allowing a PC that Steam is running on to use your steam account by sending a code to the email account associated with your steam account. Once authorized that PC can login at any time with your steam account (still requiring password). This prevents any old PC around the world from trying to log in to your Steam account without you knowing about it. The only down side is that if the PC is compromised and your steam login details are known a hacker could login to your steam account from that PC.

Steam Authenticator is better for security in that you are required to authorize the PC EVERY time the steam account is logged in to. This is great for security and normal Steam Clients but rubbish for game servers.

TL;DR Steam Guard via email is the only workable option for game servers because of automation.

Hope that makes more sense :+1:

**Comment by dgibbs64** [comment]
@UltimateByte i see in issue #570 you talk about unset the steam password. Does this cause 2-factor authentication to work? does this mean that not specifying the password allows the authenticator to only be used once like steam guard via email?

If so there may be a way to sort this....To get this to work we require 2-factor authentication code to be entered only once. Failing this I would be correct in stating that Steam Guard via email is the only option.

**Comment by Dunkelzhan** [comment]
FYI -

I tried your suggestion of removing the Steam Guard Authenticator, and using Steam Guard Email codes in an effort to fix issue #570. I deleted the "sdtduser" account on my box, and all associated files.  I then rebooted, created the user account again, and proceeded with a clean install.

The server installed correctly
 - when asked for a password, I entered mine as normal and it accepted it.
 
 - when it asked me for a Steam Guard code, i received one via email, which it accepted and completed the install.  

I then validated, and updated (it confirmed I was running the latest version, and no updates were available).  I was able to run a backup, and verified the server autostarts with the cron job.  I then tried a manual update again, as this is where I had problems in the past.  I again received a failure:

sdtdserver@WolfeDen1:~$ ./sdtdserver update
Redirecting stderr to '/home/sdtdserver/Steam/logs/stderr.txt'
[  0%] Checking for available updates...
[----] Verifying installation...
Steam Console Client (c) Valve Corporation
-- type 'quit' to exit --
Loading Steam API...OK.

Logging in user '(REMOVED USERNAME)' to Steam Public...
Using cached credentials. . .
Logged in OK
Waiting for license info...OK
[ FAIL ] Update sdtd-server: Checking for update: SteamCMD: Not returning version info

sdtdserver@WolfeDen1:~$ ./sdtdserver update
Redirecting stderr to '/home/sdtdserver/Steam/logs/stderr.txt'
[  0%] Checking for available updates...
[----] Verifying installation...
Steam Console Client (c) Valve Corporation
-- type 'quit' to exit --
Loading Steam API...OK.

Logging in user '(REMOVED USERNAME)' to Steam Public...
Using cached credentials. . .
Login Failure: Invalid Password 
FAILED with result code 5
Failure! Error running SteamCMD
[ FAIL ] Update sdtd-server: Checking for update: SteamCMD: Not returning version info
sdtdserver@WolfeDen1:~$ 

Notice on the first ./sdtdserver update, it uses the cached credentials and says 'Logged in OK', then immediately running it again it comes back with 'Login Failure: invalid Password'.

I wish I knew more to help, but wanted to provide this information if it helps you with the issue.

**Comment by dgibbs64** [comment]
I was receiving this issue after trying out Steam Guard via Smartphone also. Its a bit of a pain in the arse. Im still working on this see what can be done...if anything

Login Failure: Invalid Password 
FAILED with result code 5

**Comment by UltimateByte** [comment]
Is there anything different between login commands at install and at update ?

**Comment by DSDV** [comment]
Phu thats realy a big problem.
Maybe for what its worth try contacting Steamsupport.

It should be in there intrest to make servverhosting as easy as possible becuase its ther infrastructure?!

**Comment by dgibbs64** [comment]
@UltimateByte did you want to assign this to yourself?

**Comment by UltimateByte** [comment]
Yep, forgot to assign this one !
BTW i'll probably need a hand at some point.

**Comment by UltimateByte** [comment]
OK, so i tried to put a steamuser to CSGO, and it seems to have the exact same behaviour.

- First i only set my steamuser. It asked for my password, and steam auth only once. Everything worked well.
- Then i set my steamuser and steampass, it asks for 2 auth codes (which were different for me), and it gets stuck after entering the second one.
- Then i removed the whole folder and set ONLY the steamuser. It asked for the password, then steam guard code, then password again, and ultimately got stuck.
- Then i removed the cleared the whole folder again, went in the user's home directory (i was into a subfolder) and removed the "Steam" and ".steam" folders (that last one you can only see with "ls -al" command), and tried again : Didn't help, still stuck at the second authentication. 

Then, i tried in manual mode.
Install steamcmd in /home/serverdev/csgoserver/steamcmd
run ./steamcmd.sh
login my_username
prompt for password
prompt for 2 factor auth
force_install_dir /home/serverdev/csgoserver
app_update 740
=> Installing

Seems like steam auth is working find in standalone, so there is definitely an issue in how LGSM tries to login.
Gonna check the code to try to find out why it happens.

**Comment by UltimateByte** [comment]
Got some news !
That same issue just happened to me while using steamcmd in manual mode !
And it was cause by... Hold on... Suspense... Already being logged-in (as anonymous) !
@dgibbs64 
The fix would probably be to make it so it only logs in once !

**Comment by UltimateByte** [comment]
I've been checking for potential sessions without a +quit, but there is none...
The issue may be that it caches credentials...

Potential functions affecting this : 
https://github.com/dgibbs64/linuxgsm/blob/master/functions/install_steamcmd.sh#L35
Or
https://github.com/dgibbs64/linuxgsm/blob/master/functions/check_steamcmd.sh#L56

**Comment by UltimateByte** [comment]
OK, i had that issue while installing bb2server
It asked for my key once, then got stuck. The next times, it was stuck at the password step.
I ended up running steamcmd manually, which isn't a satisfying solution...

**Comment by Dunkelzhan** [comment]
@UltimateByte : Could you post what the commands are to update Steam CMD manually for novices like me?

**Comment by UltimateByte** [comment]
Sure !
I now know it by heart except the link ! :o))
So in a new user : 
mkdir steamcmd
cd steamcmd
wget https://steamcdn-a.akamaihd.net/client/installer/steamcmd_linux.tar.gz
tar -xvzf steamcmd_linux.tar.gz
rm steamcmd_linux.tar.gz
./steamcmd.sh
login YourUserame
force_install_dir /home/username/serverfiles
app_update 740 validate
(740 is for cs go)

BTW you can also find those instructions here
https://developer.valvesoftware.com/wiki/SteamCMD#Linux

But the whole purpose of LGSM is not having to do all this. But for games requiring login for now, it may be necessary to do it manually waiting for a fix, either from Valve or LGSM.

**Comment by Dunkelzhan** [comment]
Thank you!

FYI, i got curious, and I tried:

cd steamcmd
steamcmd.sh

then, inside the steamcmd prompt:

login (removed username) (removed password)

Steam CMD then queried for the SteamGuard key, note: I used Steam Authenticator w/ my phone for this.  I entered the code and it accepted it.

app_update 294420

This seemed to download and install updates.  I then exited steamcmd:

quit

and changed directory back to my sdtdserver install

cd ..

and ran 

./sdtdserver update

It appeared to login correctly, no issues with "cached credentials" as I had been getting, then check for updates - it found it was running the current version, where it hadn't been before. So it looks like running the update "manually" this way in Steamcmd also works. 

Just FYI in case this helps you resolve anything.

**Comment by UltimateByte** [comment]
Yest, that helps to allow us reproducing the steps !
I guess we will go deeper into that with Daniel as soon as we can.

**Comment by UltimateByte** [comment]
@dgibbs64 I messed around quite a bit with it without finding any useful thing. I guess we could investigate that together.

**Comment by DSDV** [comment]
BTW there has to be an automated way entering the mobile authentication!

how is it otherwise possible that there are thousands of trading bots out there able to trade without the 72 hold of penalty ?

**Comment by UltimateByte** [comment]
@DSDV  If you find the good way, feel free to tell us ! ^^

**Comment by dgibbs64** [comment]
Probably not using SteamCMD and using the Steam Client

**Comment by DaMaxx** [comment]
Got the Steam Guard running with this patch #893 
See #892 for informations about my system

**Comment by UltimateByte** [comment]
@DaMaxx Good to know, thanks, i just asked Cedar and Dgibbs if they knew more about unbuffer causing issues with this.

**Comment by DaMaxx** [comment]
@UltimateByte I just had another look at this issue, when I force the program to use "stdbuf -i0 -o0 -e0" instead of unbuffer it works as well, at least on my machine. 
So my guess is the problem comes with the "unbuffer" command.

**Comment by UltimateByte** [comment]
I guess @cedarlug already suggested that.
